diff --git a/extern/bullet2/src/BulletCollision/CollisionShapes/btCapsuleShape.cpp b/extern/bullet2/src/BulletCollision/CollisionShapes/btCapsuleShape.cpp
index 864df26e931..f5e94fe8cca 100644
--- a/extern/bullet2/src/BulletCollision/CollisionShapes/btCapsuleShape.cpp
+++ b/extern/bullet2/src/BulletCollision/CollisionShapes/btCapsuleShape.cpp
@@ -133,7 +133,7 @@ void	btCapsuleShape::calculateLocalInertia(btScalar mass,btVector3& inertia) con
 	btVector3 halfExtents(radius,radius,radius);
 	halfExtents[getUpAxis()]+=getHalfHeight();
 
-	btScalar margin = CONVEX_DISTANCE_MARGIN;
+	btScalar margin = getMargin();
 
 	btScalar lx=btScalar(2.)*(halfExtents[0]+margin);
 	btScalar ly=btScalar(2.)*(halfExtents[1]+margin);
diff --git a/extern/bullet2/src/BulletCollision/CollisionShapes/btConeShape.h b/extern/bullet2/src/BulletCollision/CollisionShapes/btConeShape.h
index 4a0df0d558b..9ded8e935e9 100644
--- a/extern/bullet2/src/BulletCollision/CollisionShapes/btConeShape.h
+++ b/extern/bullet2/src/BulletCollision/CollisionShapes/btConeShape.h
@@ -49,15 +49,13 @@ public:
 		btTransform identity;
 		identity.setIdentity();
 		btVector3 aabbMin,aabbMax;
-		getAabb(identity,aabbMin,aabbMax);
 
+		getAabb(identity,aabbMin,aabbMax); // This already contains the margin
 		btVector3 halfExtents = (aabbMax-aabbMin)*btScalar(0.5);
 
-		btScalar margin = getMargin();
-
-		btScalar lx=btScalar(2.)*(halfExtents.x()+margin);
-		btScalar ly=btScalar(2.)*(halfExtents.y()+margin);
-		btScalar lz=btScalar(2.)*(halfExtents.z()+margin);
+		btScalar lx=btScalar(2.)*(halfExtents.x());
+		btScalar ly=btScalar(2.)*(halfExtents.y());
+		btScalar lz=btScalar(2.)*(halfExtents.z());
 		const btScalar x2 = lx*lx;
 		const btScalar y2 = ly*ly;
 		const btScalar z2 = lz*lz;
diff --git a/extern/bullet2/src/BulletCollision/CollisionShapes/btConvexInternalShape.cpp b/extern/bullet2/src/BulletCollision/CollisionShapes/btConvexInternalShape.cpp
index 083d60b1b1e..707c5fc8c7c 100644
--- a/extern/bullet2/src/BulletCollision/CollisionShapes/btConvexInternalShape.cpp
+++ b/extern/bullet2/src/BulletCollision/CollisionShapes/btConvexInternalShape.cpp
@@ -131,8 +131,8 @@ void	btConvexInternalAabbCachingShape::recalcLocalAabb()
 	
 	for ( int i = 0; i < 3; ++i )
 	{
-		m_localAabbMax[i] = _supporting[i][i] + m_collisionMargin;
-		m_localAabbMin[i] = _supporting[i + 3][i] - m_collisionMargin;
+		m_localAabbMax[i] = _supporting[i][i];
+		m_localAabbMin[i] = _supporting[i + 3][i];
 	}
 	
 	#else
@@ -142,10 +142,10 @@ void	btConvexInternalAabbCachingShape::recalcLocalAabb()
 		btVector3 vec(btScalar(0.),btScalar(0.),btScalar(0.));
 		vec[i] = btScalar(1.);
 		btVector3 tmp = localGetSupportingVertex(vec);
-		m_localAabbMax[i] = tmp[i]+m_collisionMargin;
+		m_localAabbMax[i] = tmp[i];
 		vec[i] = btScalar(-1.);
 		tmp = localGetSupportingVertex(vec);
-		m_localAabbMin[i] = tmp[i]-m_collisionMargin;
+		m_localAabbMin[i] = tmp[i];
 	}
 	#endif
 }
